Backend/Remora.Discord.Gateway/Transport/WebSocketPayloadTransportService.cs

```diff
@@ -113,56 +140,53 @@ public async Task<Result> ConnectAsync(Uri endpoint, CancellationToken ct = defa
     /// <inheritdoc />
     public async Task<Result> SendPayloadAsync(IPayload payload, CancellationToken ct = default)
     {
-        if (_clientWebSocket is null)
+        if (_clientWebSocket?.State is not WebSocketState.Open)
         {
             return new InvalidOperationError("The transport service is not connected.");
         }
 
-        if (_clientWebSocket.State != WebSocketState.Open)
-        {
-            return new InvalidOperationError("The socket was not open.");
-        }
+        var semaphoreEntered = false;
 
-        await using var memoryStream = new MemoryStream();
-
-        byte[]? buffer = null;
         try
         {
-            await JsonSerializer.SerializeAsync(memoryStream, payload, _jsonOptions, ct);
-
-            if (memoryStream.Length > 4096)
+            semaphoreEntered = await _sendSemaphore.WaitAsync(100, ct);
```

Comment:
Interesting use of WaitAsync.

Reply:
In what way? With the current use of the transport service, the send method will never be called from more than two threads. My line of thinking is hence, if it has taken that long to enter the semaphore, something has probably gone wrong. This is certainly expecting things to be a bit more domain specific than the accessibility of the service might otherwise indicate.